lib/Coffer/Directory.hs

```diff
@@ -0,0 +1,94 @@
+module Coffer.Directory
+  ( Directory(..)
+  , emptyDir
+  , singleton
+  , allEntries
+  , mapMaybe
+  , mapEntry
+  , mapDir
+  , insertEntry
+  , trimEmptyDirs
+  , filterEntries
+  ) where
+
+import Control.Lens
+import qualified Data.HashMap.Strict as HashMap
+import Data.HashMap.Strict (HashMap)
+import qualified Data.Maybe as Maybe
+import Coffer.Path (PathSegment, entryPathParentDir, unPath)
+import Entry (Entry)
+import qualified Entry as E
+
+----------------------------------------------------------------------------
+-- Directory
+----------------------------------------------------------------------------
+
+data Directory = Directory
+  { _entries :: [Entry]
+  , _subdirs :: HashMap PathSegment Directory
+  }
+  deriving stock (Show, Eq)
+
+makeLenses 'Directory
+
+emptyDir :: Directory
+emptyDir = Directory mempty mempty
+
+singleton :: Entry -> Directory
+singleton entry = insertEntry entry emptyDir
+
+allEntries :: Directory -> [Entry]
+allEntries (Directory entries subdirs) =
+  entries <> foldMap allEntries subdirs
```

Comment:
It would be good having a function, which lists entries **only** in a specific directory.

Reply:
There was no need for that function to implement these CLI commands. We can add that function when a need for it arises.